fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,451 @@
+from typing import Any, Callable, List, Union
+
+import numpy as np
+from pylatex import Command, Document, Itemize, MultiColumn, NoEscape, Package, Section, Subsection, Table, Tabularx, \
+    escape_latex
+
+import fastestimator as fe
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.latex_util import TabularCell, WrapText
+from fastestimator.util.traceability_util import traceable
+from fastestimator.util.util import to_list, to_number
+
+
+@traceable()
+class TestCase():
+    """This class defines the test case that ModelEval trace will take to perform auto-testing.
+
+    Args:
+        description: A test description.
+        criteria: Function to perform the test returns True when the test passes and False when it fails. Input
+            variable names will be used as input keys to further derive input value.
+        aggregate: If True, this test is aggregate type and its criteria function will be examined at epoch_end. If
+            False, this test is per-instance type and its test criteria function will be examined at batch_end.
+        fail_threshold: Thershold of failure instance number to judge per-instance test as failed or passed. If failure
+            number is above this value, then the test fails; otherwise it passes. It only has effect when aggregate is
+            equal to False.
+    """
+    def __init__(self, description: str, criteria: Callable, aggregate: bool = True, fail_threshold: int = 0) -> None:
+        self.description = description
+        self.criteria = criteria
+        self.criteria_inputs = inspect.signature(criteria).parameters.keys()
+        self.aggregate = aggregate
+        if not self.aggregate:
+            self.fail_threshold = fail_threshold
+
+    def clean_result(self) -> None:
+        if not self.aggregate:
+            self.result = []
+            self.fail_id = []
+
+
+@traceable()
+class ModelEval(Trace):
+    """Automate testing and report generation.
+
+    This trace will examine all input TestCases in the test mode and generate a PDF report and a JSON test result.
+
+
+    Args:
+        test_cases: TestCase object or list of TestCase objects.
+        save_path: Where to save the output directory.
+        test_title: Title of the test.
+        data_id: Data instance ID key. If provided, then per-instance test will return failure instance ID.
+    """
+    def __init__(self,
+                 test_cases: Union[TestCase, List[TestCase]],
+                 save_path: str,
+                 test_title: str = "Test",
+                 data_id: str = None) -> None:
+
+        self.test_title = test_title
+        self.test_cases = to_list(test_cases)
+        self.instance_cases = []
+        self.aggregate_cases = []
+        self.data_id = data_id
+
+        all_inputs = set()
+        for case in self.test_cases:
+            all_inputs.update(case.criteria_inputs)
+            case.clean_result()
```

Comment:
if you fix their init method you won't need to call this here

Reply:
but I am afraid people will use the same TestCase instance to run other test experiments.